lib/mlir/Dialect/Clift/IR/CliftOps.cpp

```diff
@@ -261,6 +271,35 @@ struct ModuleValidator {
                                << " must be directly nested within a "
                                   "ModuleOp.";
 
+    if (auto Return = mlir::dyn_cast<ReturnOp>(Op)) {
+      ValueType ReturnType = {};
+
+      if (Region &R = Return.getResult(); not R.empty())
+        ReturnType = getExpressionType(R);
+
+      if (ReturnType and isVoid(FunctionReturnType))
+        return Op->emitOpError() << Op->getName()
+                                 << " cannot return expression in function "
+                                    "returning void.";
+
+      if (ReturnType != FunctionReturnType)
+        return Op->emitOpError() << Op->getName()
+                                 << " type does not match the function return "
+                                    "type";
+    } else if (mlir::isa<SwitchBreakOp>(Op)) {
+      if (not hasParentStatement(Op, ControlStatementKind::Switch))
+        return Op->emitOpError()
+               << Op->getName() << " must be nested within a switch operation.";
+    } else if (mlir::isa<LoopBreakOp>(Op)) {
+      if (not hasParentStatement(Op, ControlStatementKind::Loop))
+        return Op->emitOpError()
+               << Op->getName() << " must be nested within a loop operation.";
+    } else if (mlir::isa<LoopContinueOp>(Op)) {
+      if (not hasParentStatement(Op, ControlStatementKind::Loop, false))
+        return Op->emitOpError()
+               << Op->getName() << " must be nested within a loop operation.";
```

Comment:
`hasParentStatement` only considers loops and switches so nesting within an if-statement has no effect.

Reply:
Eh ok. But then `hasParentStatement` does not do what it advertises on the surface. Adding the comment definitely helps. But there's another issue that I didn't really brought up but that is important: I know we're going to need something like `hasParentStatement` in the future for working not just on `switch`es and loops. And `hasParentStatement` would be a perfect name for that, and it also shares 99% of the implementation. It's just that at the moment we're kind-of-artificially restricting it to only be able to reason about `switch`es and loops.

I think that we should have a `hasParentStatement` should be capable of handling any statement, and have 2 modes of operations: one-step (only checks direct parent, return `false` if the direct parent does not exist), transitive (checks all parents, to se if finds at some point one of the requested kind.

Then, the specific checks you need to perform for verifying the well-nestedness of `break`s inside loops of `switch`es can be expressed using the new `hasParentStatement`.

Reply:
> I know we're going to need something like `hasParentStatement` in the future

I would rather deal with that when those concrete use cases actually materialise and the requirements for such an interface become clearer. Since this function is very local to the module validator, would you be okay with just renaming the functions and enum along the lines of `hasLoopOrSwitchParent`?

Reply:
> I know we're going to need something like hasParentStatement in the future

I would rather deal with that when those concrete use cases actually materialise and the requirements for such an interface become clearer. Since this function is very local to the module validator, would you be okay with just renaming the functions and enum along the lines of `hasLoopOrSwitchParent`?

Reply:
Ok that's fine.